Research/Research/RSDTaskViewModel.swift

```diff
@@ -0,0 +1,947 @@
+        var path: RSDPathComponent? = self
+        while let nextPath = path {
+            if let taskPath = path as? RSDTaskPathComponent,
+                let task = taskPath.task,
+                let thisStep = taskPath.currentChild?.step,
+                task.stepNavigator.hasStep(after: thisStep, with: nextPath.taskResult) {
+                // If the current task path has more steps then this is not the end
+                return true
+            }
+            path = nextPath.parent
+        }
+        
+        // Otherwise, this is the last step
+        return false
+    }
+    
+    /// Is there previous step that this task can go back to?
+    public var hasStepBefore: Bool {
+        // Exit early if this is the first step. There is no back button.
+        if self.currentStep == nil && self.parent == nil {
+            return false
+        }
+        
+        // While this is the first step, look up the path chain for a task that has not reached the end
+        var path: RSDPathComponent? = self
+        while let nextPath = path {
+            if let taskPath = path as? RSDTaskPathComponent,
+                let task = taskPath.task,
+                let thisStep = taskPath.currentChild?.step,
+                task.stepNavigator.hasStep(before: thisStep, with: nextPath.taskResult) {
+                // If the current task path has more steps then this is not the end
+                return true
+            }
+            path = nextPath.parent
+        }
+        
+        // Otherwise, this is the last step
+        return false
+    }
+    
+    /// Go forward to the next step. If the task is not loaded for the current point in the task path, then
+    /// it will attempt to fetch it again.
+    open func goForward() {
+        guard let viewModel = self.currentTaskPath as? RSDTaskViewModel, viewModel == self
+            else {
+                self.currentTaskPath.moveForwardToNextStep()
+                return
+        }
+        _goForward()
+    }
+    
+    private func _goForward() {
+        
+        guard self.task != nil else {
+            _fetchTaskFromCurrentInfo()
+            return
+        }
+        
+        // Add or update the result end date if there was previous step
+        if let previousStep = self.currentChild {
+            if let result = self.taskResult.stepHistory.last, result.identifier == previousStep.identifier {
+                var finalResult = result
+                finalResult.endDate = Date()
+                self.taskResult.appendStepHistory(with: finalResult)
+            }
+            else {
+                let result = previousStep.pathResult()
+                self.taskResult.appendStepHistory(with: result)
+            }
+        }
+        
+        // move to the next step
+        _moveToNextStep()
+    }
+    
+    /// Go back to the previous step.
+    ///
+    /// - note: This method will throw an assertion if there isn't a previous step.
```

Comment:
looks like it throws if there isn't a *current* step, or am I confused?

Reply:
Good point, if this isn't the top level path then it should go back to the previous parent path (which it does) and should not ever assert.